### ISSUE TITLE
[Bug]: Never uses preprocessors

### ISSUE THREAD

**Issue by narukaze132** [Mikubill/sd-webui-controlnet#1651]
### Is there an existing issue for this?

- [X] I have searched the existing issues and checked the recent builds/commits of both this extension and the webui

### What happened?

Whenever I generate an image with ControlNet active, the console logs say that it's using the "none" preprocessor, and as far as I can tell it is indeed using that preprocessor. This happens regardless of what preprocessor I have selected, tested with openpose, hed, and several others.

### Steps to reproduce the problem

1. Go to the ControlNet configuration in WebUI
2. Upload an image to use as input
3. Select any preprocessor that isn't the "none" preprocessor
5. Enable the node if you haven't already
6. Select a ControlNet model
7. Generate any image
8. Wait until the generation completes
9. Check the WebUI logs, it should say "Loading preprocessor: none"
10. Try drawing on the input image using inpainting and doing it again if you want, you should get the same message in the log

### What should have happened?

The preprocessor I selected should have been applied to the input image, rather than it just using the "none" preprocessor.

### Commit where the problem happens

webui: Version 1.3.2, the current release version
controlnet: [c9c7317](https://github.com/Mikubill/sd-webui-controlnet/commit/c9c7317cea607ea43cfa22139e0e2bf606cba027)

### What browsers do you use to access the UI ?

Mozilla Firefox

### Command Line Arguments

```Shell
--medvram --xformers
```

### List of enabled extensions

a1111-sd-webui-lycoris
sd-webui-controlnet
LDSR
Lora
ScuNET
SwinIR
prompt-bracket-checker

### Console logs

```Shell
venv "D:\Programs\stable-diffusion-webui\venv\Scripts\Python.exe"
Python 3.10.6 (tags/v3.10.6:9c7b4bd, Aug  1 2022, 21:53:49) [MSC v.1932 64 bit (AMD64)]
Version: v1.3.2
Commit hash: baf6946e06249c5af9851c60171692c44ef633e0
Installing requirements

Launching Web UI with arguments: --medvram --xformers --no-gradio-queue
2023-06-17 13:26:04,259 - ControlNet - INFO - ControlNet v1.1.224
ControlNet preprocessor location: D:\Programs\stable-diffusion-webui\extensions\sd-webui-controlnet\annotator\downloads
2023-06-17 13:26:04,791 - ControlNet - INFO - ControlNet v1.1.224
Loading weights [a5d0b8e961] from D:\Programs\stable-diffusion-webui\models\Stable-diffusion\nai.safetensors
Running on local URL:  http://127.0.0.1:7860

To create a public link, set `share=True` in `launch()`.
Startup time: 22.8s (import torch: 5.8s, import gradio: 4.0s, import ldm: 1.5s, other imports: 4.6s, setup codeformer: 0.4s, list SD models: 0.3s, load scripts: 3.9s, initialize extra networks: 0.1s, scripts before_ui_callback: 0.2s, create ui: 0.9s, gradio launch: 1.2s).
Creating model from config: D:\Programs\stable-diffusion-webui\models\Stable-diffusion\nai.yaml
LatentDiffusion: Running in eps-prediction mode
DiffusionWrapper has 859.52 M params.
Loading VAE weights found near the checkpoint: D:\Programs\stable-diffusion-webui\models\Stable-diffusion\nai.vae.pt
Applying optimization: xformers... done.
Textual inversion embeddings loaded(4): badhandv4, easynegative, hiten-style-nao, hiten1
Model loaded in 31.4s (load weights from disk: 3.1s, create model: 0.6s, apply weights to model: 21.7s, apply half(): 0.9s, load VAE: 5.0s).
2023-06-17 13:29:50,000 - ControlNet - INFO - Loading model: control_sd15_hed [fef5e48e]
2023-06-17 13:29:52,715 - ControlNet - INFO - Loaded state_dict from [D:\Programs\stable-diffusion-webui\extensions\sd-webui-controlnet\models\control_sd15_hed.pth]
2023-06-17 13:29:52,715 - ControlNet - INFO - Loading config: D:\Programs\stable-diffusion-webui\extensions\sd-webui-controlnet\models\control_sd15_hed.yaml
2023-06-17 13:29:56,144 - ControlNet - INFO - ControlNet model control_sd15_hed [fef5e48e] loaded.
2023-06-17 13:29:56,722 - ControlNet - INFO - using mask as input
2023-06-17 13:29:56,724 - ControlNet - INFO - API used default value: unit.threshold_a = 0
2023-06-17 13:29:56,724 - ControlNet - INFO - API used default value: unit.threshold_b = 0
[NOTE: I had softedge_hed selected in ControlNet at this point]
2023-06-17 13:29:56,724 - ControlNet - INFO - Loading preprocessor: none
2023-06-17 13:29:56,724 - ControlNet - INFO - preprocessor resolution = 512
Total progress: 100%|##########| 20/20 [00:37<00:00,  1.83s/it]
```

### Additional information

I got the provided console log through tee (or rather the PowerShell equivalent), so the escape codes didn't work correctly. I've fixed up the problems, but I felt I should note they were there.

Speaking of the logs, I had the "softedge_hed" preprocessor selected when I generated that image.

**Comment by huchenlei** [comment]
`2023-06-17 13:29:56,722 - ControlNet - INFO - using mask as input` indicates that you have a inpaint mask on the controlnet input image. In this case, preprocessor will not be run, and controlnet will directly use the mask as input.

See #1638 for more details. 

We are debating whether we should change this behaviour on non-inpaint models. You can join the discussion in #1638. Closing this issue as duplicated.

**Comment by narukaze132** [comment]
The thing is, there shouldn't have been an inpaint mask on the image. I specifically removed the alpha channel before uploading the image, and didn't click on the input image at all after uploading it, just to try and avoid this exact issue. Yet somehow it still registered as having an inpaint mask, and hence didn't get processed properly.

**Comment by huchenlei** [comment]
If you click the run preprocessor button, what is on the preview image?

**Comment by narukaze132** [comment]
Nothing. It's completely black, no matter what preprocessor I use. (Except the invert preprocessor, but that just makes it completely white instead.)

**Comment by huchenlei** [comment]
I don't think it would be completely black or white and still triggers the `have_mask` logic:

`have_mask = 'mask' in image and not ((image['mask'][:, :, 0] == 0).all() or (image['mask'][:, :, 0] == 255).all())`

You can download the generated image, convert it to numpy array and check if all pixels are black/white.

**Comment by narukaze132** [comment]
I downloaded the preprocessed image, used cv2 to load it into Python, and checked if any of the values were not 0. It came back false.

I ended up adding a line to the code that printed out the image's mask just to check, and it had a few 1s in it, which fooled the check. The thing is, it _shouldn't_ have had those 1s, because I hadn't clicked on the image at all. So I'm 99% sure something's bugged, here.

**Comment by cian0** [comment]
Same thing happening here, no mask uploaded there, txt2img doesn't work with image as the reference in controlnet tab. Img2img works just fine

**Comment by cian0** [comment]
<img width="578" alt="image" src="https://github.com/Mikubill/sd-webui-controlnet/assets/1121189/76a8abcc-d429-4627-a502-ee8b2e9b15f9">
<img width="773" alt="image" src="https://github.com/Mikubill/sd-webui-controlnet/assets/1121189/f825f7c2-08ea-4474-a86d-a3e1f5427fe5">

**Comment by huchenlei** [comment]
It came into my attention that even I do not have any paint on the image, I can still get a mask in the parameter passed to ControlNet. I got something like:
```
data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAARMAAAC3CAIAAAC+MS2jAAAAqUlEQVR4nO3BAQ0AAADCoPdPbQ8HFAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA/wZOlAAB5tU+nAAAAABJRU5ErkJggg==
```
as mask. 

However, I verified with script that all pixels are black, so essentially it has no effect. Also, the blank mask is not always attached to the image. In some runs I get it, in others I do not. 

You can use this extension to verify the parameters passed to ControlNet:
https://github.com/huchenlei/sd-webui-api-payload-display

**Comment by AlanZhou2022** [comment]
i have the same problem, without any fix so far.

**Comment by Daimonin** [comment]
Sounding in, same issue. No mater what I do, controlnet finds a non-existent mask and refuses to use the preprocess.

2023-08-12 14:11:40,694 - ControlNet - INFO - Loading model from cache: controlnet11Models_openpose [73c2b67d]
2023-08-12 14:11:40,696 - ControlNet - INFO - using mask as input
2023-08-12 14:11:40,702 - ControlNet - INFO - Loading preprocessor: none
2023-08-12 14:11:40,702 - ControlNet - INFO - preprocessor resolution = 512

If the system to detect a mask is bugged, could we get at least a temporary fix in the form of a toggle to turn mask on/off manually? Or anyone know when the bug popped up to begin with?

**Comment by Daimonin** [comment]
For anyone else lost and wandering, hoping some solution is found when there's been no word on what's even causing this problem in two months.

In the meantime, you can continue to use ControlNet by opening up scripts/controlnet.py

Comment out:
have_mask = 'mask' in image and not ((image['mask'][:, :, 0] == 0).all() or (image['mask'][:, :, 0] == 255).all())
And add:
have_mask = False

It's hacky and means you won't be able to use masks with controlnet, but at least you'll be able to USE controlnet at all!

**Comment by huchenlei** [comment]
I will add a flag in the settings to ban the mask usage in non-inpaint control types.

**Comment by narukaze132** [comment]
I ended up "fixing" it using an approach similar to Daimonin's, but slightly more nuanced. Instead of doing `have_mask = False`, I changed the have_mask line to this:

`have_mask = 'mask' in image and not ((image['mask'][:, :, 0] <= 5).all() or (image['mask'][:, :, 0] >= 250).all())`

Basically, the bug comes from the mask somehow having some 1s in it, despite, y'know, not *existing*. The workaround adds some "fuzziness" to the check, so if none of the pixels have a value of at least 5, the mask doesn't register.

The only issue I've found with this fix is that the preview still acts like there's a mask on the image (and so is completely black), but when I actually tried using it in generation, it worked just fine.

**Comment by Daimonin** [comment]
@narukaze132 Nice! I've been out of the coding scene in over a decade so I'll try your less brute force approach.

The reason the preview is still showing a mask is because the controlnet looks to use different code to generate the preview. I found the 'have_mask' line by literally searching the files for the "using mask as input" string. Preview does NOT generate that string in console, hence, there's another spot in the code that would need to have the fuzzy check applied for the preview.

@huchenlei That will be much appreciated!
